tensorflow_lite_support/ios/task/audio/core/audio_record/sources/TFLAudioRecord.m

```diff
@@ -0,0 +1,269 @@
+    return pcmBuffer;
+  };
+
+  NSError *conversionError = nil;
+  AVAudioConverterOutputStatus converterStatus = [audioConverter convertToBuffer:outPCMBuffer
+                                                                           error:&conversionError
+                                                              withInputFromBlock:inputBlock];
+
+  switch (converterStatus) {
+    case AVAudioConverterOutputStatus_HaveData: {
+      return outPCMBuffer;
+    }
+    case AVAudioConverterOutputStatus_Error: {
+      NSString *errorDescription = conversionError.localizedDescription
+                                       ? conversionError.localizedDescription
+                                       : @"Some error occured while processing incoming audio "
+                                         @"frames.";
+      [TFLCommonUtils createCustomError:error
+                             withDomain:TFLAudioRecordErrorDomain
+                                   code:TFLAudioRecordErrorCodeProcessingError
+                            description:errorDescription];
+      break;
+    }
+    case AVAudioConverterOutputStatus_EndOfStream: {
+      [TFLCommonUtils createCustomError:error
+                             withDomain:TFLAudioRecordErrorDomain
+                                   code:TFLAudioRecordErrorCodeProcessingError
+                            description:@"Reached end of input audio stream."];
+      break;
+    }
+    case AVAudioConverterOutputStatus_InputRanDry: {
+      [TFLCommonUtils createCustomError:error
+                             withDomain:TFLAudioRecordErrorDomain
+                                   code:TFLAudioRecordErrorCodeProcessingError
+                            description:@"Not enough input is available to satisfy the request."];
+      break;
+    }
+  }
+  return nil;
+}
+
+- (BOOL)loadAudioPCMBuffer:(AVAudioPCMBuffer *)pcmBuffer error:(NSError **)error {
+  if (pcmBuffer.frameLength == 0) {
+    [TFLCommonUtils createCustomError:error
+                           withDomain:TFLAudioRecordErrorDomain
+                                 code:TFLAudioRecordErrorCodeInvalidArgumentError
+                          description:@"You may have to try with a different "
+                                      @"channel count or sample rate"];
+  } else if ((pcmBuffer.frameLength % self.audioFormat.channelCount) != 0) {
+    [TFLCommonUtils createCustomError:error
+                           withDomain:TFLAudioRecordErrorDomain
+                                 code:TFLAudioRecordErrorCodeInvalidArgumentError
+                          description:@"You have passed an unsupported number of channels."];
+  } else {
+    TFLFloatBuffer *floatBuffer = [[TFLFloatBuffer alloc] initWithData:pcmBuffer.floatChannelData[0]
+                                                                  size:pcmBuffer.frameLength];
+
+    if ([self->_ringBuffer loadBuffer:floatBuffer offset:0 size:floatBuffer.size error:error]) {
```

Comment:
You're doing memory copy twice here: from pcmBuffer.floatChannelData to floatBuffer then to _ringBuffer, which may cause added latency. This can be reduced to one copy by adding the `loadFloat:(float *)` API to the TFLRingBuffer.

Reply:
I have changed the interface of TFLRingBuffer. Load can be done using `loadFloatData:(float *) dataSize:(NSUInteger)`. I have removed the load from `TFLFloatBuffer` as it won't be used by any of our APIs.  `TFLFloatBuffer` will only be used for returning the data using` readAtOffset`.